aemcmc/gibbs.py

```diff
@@ -13,52 +18,143 @@
 )
 
 
-def update_beta_low_dimension(rng, omega, lambda2tau2_inv, X, z):
+def update_beta_low_dimension(
+    srng: RandomStream,
+    omega: TensorVariable,
+    lmbdatau_inv: TensorVariable,
+    X: TensorVariable,
+    z: TensorVariable,
+) -> TensorVariable:
     Q = X.T @ (omega[:, None] * X)
     indices = at.arange(Q.shape[1])
     Q = at.subtensor.set_subtensor(
         Q[indices, indices],
-        at.diag(Q) + lambda2tau2_inv,
+        at.diag(Q) + lmbdatau_inv,
     )
-    return multivariate_normal_rue2005(rng, X.T @ (omega * z), Q)
+    return multivariate_normal_rue2005(srng, X.T @ (omega * z), Q)
 
 
-def update_beta_high_dimension(rng, omega, lambda2tau2_inv, X, z):
-    return multivariate_normal_cong2017(rng, lambda2tau2_inv, omega, X, z)
+def update_beta_high_dimension(
+    srng: RandomStream,
+    omega: TensorVariable,
+    lmbdatau_inv: TensorVariable,
+    X: TensorVariable,
+    z: TensorVariable,
+) -> TensorVariable:
+    return multivariate_normal_cong2017(srng, lmbdatau_inv, omega, X, z)
 
 
-def update_beta(rng, omega, lambda2tau2_inv, X, z):
+def update_beta(
+    srng: RandomStream,
+    omega: TensorVariable,
+    lmbdatau_inv: TensorVariable,
+    X: TensorVariable,
+    z: TensorVariable,
+) -> TensorVariable:
     return ifelse(
         X.shape[1] > X.shape[0],
-        update_beta_high_dimension(rng, omega, lambda2tau2_inv, X, z),
-        update_beta_low_dimension(rng, omega, lambda2tau2_inv, X, z),
+        update_beta_high_dimension(srng, omega, lmbdatau_inv, X, z),
+        update_beta_low_dimension(srng, omega, lmbdatau_inv, X, z),
     )
 
 
+# horseshoe ----------------------------------------------------------
+
+halfcauchy_1_lv, halfcauchy_2_lv = var(), var()
+zero_lv = var()
+horseshoe_pattern = etuple(
+    etuplize(at.random.normal),
+    var(),
+    var(),
+    var(),
+    zero_lv,
+    etuple(etuplize(at.mul), halfcauchy_1_lv, halfcauchy_2_lv),
+)
+
+
+def horseshoe_model(srng: TensorVariable) -> TensorVariable:
+    """Horseshoe shrinkage prior [1]_.
+
+    References
+    ----------
+    .. [1]: Piironen, J., & Vehtari, A. (2017). Sparsity information
+            and regularization in the horseshoe and other shrinkage priors.
+            Electronic Journal of Statistics, 11(2), 5018-5051.
```

Comment:
Let's reference the original HS paper instead.

Reply:
Good idea, I'll look for it in the references of this one